pkg/cache/scheduler_callback_test.go

```diff
@@ -0,0 +1,626 @@
+}
+
+func TestUpdateApplication_Rejected(t *testing.T) {
+	callback, context := initCallbackTest(t, false, false)
+	defer dispatcher.UnregisterAllEventHandlers()
+	defer dispatcher.Stop()
+	app := context.getApplication(appID)
+	app.sm.SetState(ApplicationStates().Submitted)
+	NewPlaceholderManager(context.apiProvider.GetAPIs())
+	recorder := k8sEvents.NewFakeRecorder(1024)
+	events.SetRecorder(recorder)
+
+	err := callback.UpdateApplication(&si.ApplicationResponse{
+		Rejected: []*si.RejectedApplication{
+			{
+				ApplicationID: appID,
+				Reason:        "test failure",
+			},
+		},
+	})
+	assert.NilError(t, err, "error updating application")
+	err = utils.WaitForCondition(func() bool {
+		return app.sm.Current() == ApplicationStates().Failed
+	}, 10*time.Millisecond, time.Second)
+	assert.NilError(t, err, "application has not transitioned to Failed state")
+	assert.Equal(t, 1, len(recorder.Events), "no K8s event received")
+	event := <-recorder.Events
+	assert.Assert(t, strings.Contains(event, "test failure"), "event does not contain 'test failure': %s", event)
+}
+
+func TestUpdateApplication_Completed(t *testing.T) {
+	callback, context := initCallbackTest(t, false, false)
+	defer dispatcher.UnregisterAllEventHandlers()
+	defer dispatcher.Stop()
+
+	err := callback.UpdateApplication(&si.ApplicationResponse{
+		Updated: []*si.UpdatedApplication{
+			{
+				ApplicationID: appID,
+				State:         ApplicationStates().Completed,
+			},
+		},
+	})
+	assert.NilError(t, err, "error updating application")
+	assert.Assert(t, context.getApplication(appID) == nil, "application was not removed")
+}
+
+func TestUpdateApplication_Resuming_AppReserving(t *testing.T) {
+	callback, context := initCallbackTest(t, false, false)
+	defer dispatcher.UnregisterAllEventHandlers()
+	defer dispatcher.Stop()
+	app := context.getApplication(appID)
+	app.sm.SetState(ApplicationStates().Reserving)
+
+	err := callback.UpdateApplication(&si.ApplicationResponse{
+		Updated: []*si.UpdatedApplication{
+			{
+				ApplicationID: appID,
+				State:         ApplicationStates().Resuming,
+			},
+		},
+	})
+	assert.NilError(t, err, "error updating application")
+	err = utils.WaitForCondition(func() bool {
+		return app.sm.Current() == ApplicationStates().Resuming
+	}, 10*time.Millisecond, time.Second)
+	assert.NilError(t, err, "application has not transitioned to Resuming state")
+}
+
+func TestUpdateApplication_Resuming_AppNotReserving(t *testing.T) {
+	callback, context := initCallbackTest(t, false, false)
+	defer dispatcher.UnregisterAllEventHandlers()
+	defer dispatcher.Stop()
+	app := context.getApplication(appID)
+	app.sm.SetState(ApplicationStates().Submitted) // not in Reserving - nothing should happen
+
+	err := callback.UpdateApplication(&si.ApplicationResponse{
+		Updated: []*si.UpdatedApplication{
+			{
+				ApplicationID: appID,
+				State:         ApplicationStates().Resuming,
+			},
+		},
+	})
+	assert.NilError(t, err, "error updating application")
```

Comment:
We can extend it with a time window that the state has indeed not changed. It's a bit difficult to properly, it would be easier with Java-like mock object (eg. `verifyZeroInteractions(context, dispatcher)`), but we can't do that here.

Reply:
Added a one-second wait + assertion.